### PR TITLE
Fix LTO + static libraries on GCC and Clang

### PR DESCRIPTION
**Pull request by nirbheek**
Tests https://github.com/mesonbuild/meson/issues/1646

### CODE REVIEW

File: mesonbuild/environment.py [hunk 1]
```diff
@@ -693,28 +695,39 @@ def detect_static_linker(self, compiler):
...
+            linkers = [linker]
         else:
             evar = 'AR'
             if evar in os.environ:
-                linker = shlex.split(os.environ[evar])
+                linkers = [shlex.split(os.environ[evar])]
             elif isinstance(compiler, VisualStudioCCompiler):
-                linker = [self.vs_static_linker]
+                linkers = [self.vs_static_linker]
+            elif isinstance(compiler, GnuCompiler):
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'GnuCompiler' may be undefined, or defined from star imports: .compilers

File: mesonbuild/environment.py [hunk 2]
```diff
@@ -693,28 +695,39 @@ def detect_static_linker(self, compiler):
...
+                # Use gcc-ar if available; needed for LTO
+                linkers = [self.gcc_static_linker, self.default_static_linker]
+            elif isinstance(compiler, ClangCompiler):
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'ClangCompiler' may be undefined, or defined from star imports: .compilers

File: mesonbuild/environment.py [hunk 3]
```diff
@@ -693,28 +695,39 @@ def detect_static_linker(self, compiler):
...
+                arg = '/?'
+            else:
+                arg = '--version'
+            try:
+                p, out, err = Popen_safe(linker + [arg])
+            except OSError as e:
+                popen_exceptions[' '.join(linker + [arg])] = e
+                continue
+            if '/OUT:' in out or '/OUT:' in err:
+                return VisualStudioLinker(linker)
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'VisualStudioLinker' may be undefined, or defined from star imports: .compilers

File: mesonbuild/environment.py [hunk 4]
```diff
@@ -693,28 +695,39 @@ def detect_static_linker(self, compiler):
...
+            if p.returncode == 0:
+                return ArLinker(linker)
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'ArLinker' may be undefined, or defined from star imports: .compilers

File: mesonbuild/environment.py [hunk 5]
```diff
@@ -693,28 +695,39 @@ def detect_static_linker(self, compiler):
...
+            if p.returncode == 1 and err.startswith('usage'): # OSX
+                return ArLinker(linker)
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'ArLinker' may be undefined, or defined from star imports: .compilers

File: mesonbuild/environment.py (hunk 1 above)
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'GnuCompiler' may be undefined, or defined from star imports: .compilers

File: mesonbuild/environment.py (hunk 2 above)
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'ClangCompiler' may be undefined, or defined from star imports: .compilers

File: mesonbuild/environment.py (hunk 3 above)
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'VisualStudioLinker' may be undefined, or defined from star imports: .compilers

File: mesonbuild/environment.py (hunk 4 above)
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'ArLinker' may be undefined, or defined from star imports: .compilers

File: mesonbuild/environment.py (hunk 5 above)
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[flake8]__
* [F405] 'ArLinker' may be undefined, or defined from star imports: .compilers

**Comment by nirbheek** [comment]
This is unlikely to cause regressions, and would help systemd's Meson build files get feature-parity with their Autotools build files, so I think we can include this in the release.

**Comment by keszybz** [comment]
This works OK with gcc.

With clang, linking fails. The reason it fails is the static libraries that are deps of a shared library, and which are filtered out by #1545, break linking. Unfortunately this PR conflicts with #1545. When I do the filtering by hand (i.e. take the failing link command, remove the `.a` args by hand, and run that), it links fine.

So from my side, +1 to this, and I hope #1545 gets merged ;)

**Comment by nirbheek** [comment]
I'm working on #1545, but it's unlikely that it'll be ready in time for the release.

**Pull request merged**